### PR TITLE
Split tests. Part 1.

### PR DESCRIPTION
**Pull request by antrix1989**
Changed tests accordingly to the following rules:
1. Only one logical assertion per test.
2. Test only one code unit at a time.
3. Follow [Arrange-Act-Assert](http://wiki.c2.com/?ArrangeActAssert) pattern.
4. Follow our test naming convention: `test[Method]_when[Condition]_should[Result]`
5. Keep test **simple**.

### CODE REVIEW

File: ADAL/tests/unit/ADAuthenticationErrorTests.m [hunk 1]
```diff
@@ -40,70 +40,88 @@ - (void)tearDown
...
 
-- (void)testErrorFromOAuthError
+#pragma mark - errorFromAuthenticationError
+
+- (void)testErrorFromAuthenticationError_whenErrorOAuthHprotocolCodeValiderrorDetailsValidCorrelationIdNil_shouldReturnError
 {
-    NSString* details = @"Some details";
-    NSString* protocolCode = @"procol code";
-    ADAuthenticationError* error = [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_SERVER_OAUTH protocolCode:protocolCode errorDetails:details correlationId:nil];
+    NSString *details = @"Some details";
```
Comment (oldalton) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: nit: this test name is a little bit on the limit of readability. Can we try to shorten it a little bit?

File: ADAL/tests/unit/ADAuthenticationErrorTests.m [hunk 2]
```diff
@@ -40,70 +40,88 @@ - (void)tearDown
...
 
-- (void)testDescription
+- (void)testErrorFromAuthenticationError_whenErrorValidprotocolCodeValiderrorDetailsValidCorrelationIdNil_shouldReturnErrorWithDescription
 {
-    NSString* details = @"Some details";
-    NSString* protocolCode = @"some-protocol-code";
-    ADAuthenticationError* error = [ADAuthenticationError errorFromAuthenticationError:42 protocolCode:protocolCode errorDetails:details correlationId:nil];
-    XCTAssertTrue([error.description containsString:details]);
-    XCTAssertTrue([error.description containsString:protocolCode]);
+    ADAuthenticationError *error = [ADAuthenticationError errorFromAuthenticationError:42 protocolCode:@"some-protocol-code" errorDetails:@"Some details" correlationId:nil];
```
Comment (oldalton) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: nit: this test name is a little bit on the limit of readability. Can we try to shorten it a little bit?

File: ADAL/tests/unit/ADAuthenticationParametersTests.m [hunk 3]
```diff
@@ -159,260 +197,620 @@ - (void)testParametersFromResourceUrl_whenHttpsResourceUrlExists_shouldReturnAut
...
-    XCTAssertNil(params);
+    ADAuthenticationError *error;
+    
+    [ADAuthenticationParameters parametersFromResponseAuthenticateHeader:@"Bearer authorization_uri=\".\\..\\windows\\system32\\drivers\\etc\\host\"" error:&error];
+    
+    XCTAssertNotNil(error);
+}
+
+- (void)testParametersFromResponseAuthenticateHeader_whenHeaderIsValid_shouldReturnNilParameters
+{
```
Comment (oldalton) [inline ADAL/tests/unit/ADAuthenticationParametersTests.m]: should it be: whenHeaderIs**In**valid?

File: ADAL/tests/unit/ADAuthenticationContextTests.m [hunk 4]
```diff
@@ -30,138 +30,160 @@
...
     context = [ADAuthenticationContext authenticationContextWithAuthority:nil error:&error];
-    XCTAssertNil(context);
+    
     XCTAssertNotNil(error);
     XCTAssertEqual(error.code, AD_ERROR_DEVELOPER_INVALID_ARGUMENT);
     ADTAssertContains(error.errorDetails, @"authority");
-    error = nil;
+}
+
+- (void)testAuthenticationContextWithAuthority_whenAuthorityNil_shouldReturnNilContext
```
Comment (RPangrle) [inline ADAL/tests/unit/ADAuthenticationContextTests.m]: This test also needs to verify that the proper error message is returned

File: ADAL/tests/unit/ADAuthenticationContextTests.m [hunk 5]
```diff
@@ -30,138 +30,160 @@
...
+}
+
+- (void)testAuthenticationContextWithAuthority_whenAuthorityNilValidatedAuthorityNo_shouldReturnError
+{
+    ADAuthenticationContext* context = nil;
+    ADAuthenticationError* error = nil;
     
     context = [ADAuthenticationContext authenticationContextWithAuthority:nil validateAuthority:NO error:&error];
-    XCTAssertNil(context);
+
```
Comment (RPangrle) [inline ADAL/tests/unit/ADAuthenticationContextTests.m]: Should also verify that no context is returned

File: ADAL/tests/unit/ADAuthenticationErrorTests.m (hunk 1 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: changed to this one: testErrorFromAuthenticationError_whenErrorValidCodeValidErrorDetailsValidCorrelationIdNil_shouldReturnErrorWithDescription Please let me know if you have better name :)

File: ADAL/tests/unit/ADAuthenticationErrorTests.m (hunk 2 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: changed to this one: testErrorFromAuthenticationError_whenErrorValidCodeValidErrorDetailsValidCorrelationIdNil_shouldReturnErrorWithDescription Please let me know if you have a better name :)

File: ADAL/tests/unit/ADAuthenticationErrorTests.m (hunk 1 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: changed to this one testErrorFromAuthenticationError_whenCodeValidErrorProtocolValidDetailsValidCorrelationIdNil_shouldReturnError

File: ADAL/tests/unit/ADAuthenticationParametersTests.m (hunk 3 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationParametersTests.m]: right, changed.

File: ADAL/tests/unit/ADAuthenticationContextTests.m (hunk 4 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationContextTests.m]: fixed

File: ADAL/tests/unit/ADAuthenticationContextTests.m (hunk 5 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationContextTests.m]: fixed

File: ADAL/tests/unit/ADAuthenticationErrorTests.m [hunk 6]
```diff
@@ -40,70 +40,88 @@ - (void)tearDown
...
+
+- (void)testNew_shouldThrow
 {
     XCTAssertThrows([ADAuthenticationError new], @"The new selector should not work due to requirement to use the parameterless init. At: '%s'", __PRETTY_FUNCTION__);
 }
 
-- (void)testParameterlessInit
+- (void)testParameterlessInit_shouldThrow
 {
-    XCTAssertThrows([ADAuthenticationError init], @"Parameterless init should throw. At: '%s'", __PRETTY_FUNCTION__);
```
Comment (RPangrle) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: How did this compile before????

Comment (antrix1989) [inline ADAL/tests/unit/ADAuthenticationErrorTests.m]: Not sure why compiler let it pass. Seems there is +init method in NSObject, but I can not find it in public headers and if it is private, complier should tell us that there is no such method. Will ask on stackoveflow, it is interesting question :). BTW, it throws "NSInvalidArgumentException" when you try to invoke [NSObject init], so it is why the test didn't fail before.

**Pull request merged**